rust/foxglove/src/mcap_writer/async_mcap_sink.rs

```diff
@@ -0,0 +1,420 @@
+//! Async MCAP sink that doesn't block on disk I/O.
+
+use std::collections::hash_map::Entry;
+use std::collections::{BTreeMap, HashMap};
+use std::io::{Seek, Write};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use crate::{get_runtime_handle, ChannelDescriptor, ChannelId, FoxgloveError, Metadata, RawChannel, Sink, SinkChannelFilter, SinkId};
+
+type McapChannelId = u16;
+
+/// Maximum number of messages that can be queued before new messages are dropped.
+const QUEUE_CAPACITY: usize = 1024;
+
+/// A queued log message.
+struct QueuedLog {
+    descriptor: ChannelDescriptor,
+    msg: Box<[u8]>,
+    metadata: Metadata,
+}
+
+/// Commands for the background writer task.
+enum WriteCommand<W> {
+    Log(QueuedLog),
+    Metadata { name: String, data: BTreeMap<String, String> },
+    Finish(tokio::sync::oneshot::Sender<Result<W, FoxgloveError>>),
+}
+
+/// Background loop that processes queued write commands.
+/// Reads from the channel and writes to the file.
+async fn run_writer_loop<W: Write + Seek>(
+    rx: flume::Receiver<WriteCommand<W>>,
+    mut writer: mcap::Writer<W>,
+) {
+    let mut channel_map: HashMap<ChannelId, Option<McapChannelId>> = HashMap::new();
+    let mut channel_seq: HashMap<McapChannelId, u32> = HashMap::new();
+
+    while let Ok(cmd) = rx.recv_async().await {
+        match cmd {
+            WriteCommand::Log(log) => {
+                if let Err(e) = write_log(&mut writer, &mut channel_map, &mut channel_seq, &log) {
+                    tracing::error!("MCAP write error: {e}");
+                }
+            }
+            WriteCommand::Metadata { name, data } => {
+                let _ = writer.write_metadata(&mcap::records::Metadata { name, metadata: data });
+            }
+            WriteCommand::Finish(done) => {
+                let result = writer.finish().map_err(FoxgloveError::from);
+                let _ = done.send(result.map(|_summary| writer.into_inner()));
+                return;
+            }
+        }
+    }
+    let _ = writer.finish();
+}
+
+fn write_log<W: Write + Seek>(
```

Comment:
There's some duplicated logic between here and the sync version's log() function that isn't specialized based on whether it's in an async context. Consider refactoring to unify.